### PR TITLE
[POC] Export chat trascripts from Twilio (date rage)

### PR DESCRIPTION
**Pull request by GPaoloni**
## Description
This PR is a POC on how to use Twilio SDK to interact with the chat channel APIs, and export chat transcripts. 
To fetch a single chat channel transcript we need `accountSid`, `serviceSid` and `channelSid` (and of course the credentials for this account).
Here we are adding:
- Logic on contacts to extract `accountSid`, `serviceSid` and `channelSid` for all the contact records created within a range of time.
- Functions that interacts with the above, and given a list of objects like `{ accountSid, serviceSid, channelSid}`, will fetch the transcription for each. Here the transcript will be placed as a json file in the local system.
- Test file to interact with and test the above.

Some thoughts after going through this (assuming this transcript export is triggered periodically, and not after each chat task ends):
- We'll need to fetch the required information from the `Contacts` table, so we'll need to send at least one request to `hrm-service` app.
- One option to interact with Twilio APIs to extract the transcript is from within Serverless. In this casewe don't need credentials but we'll need to keep track of "which is the Serverless env for each account". This can be done in env vars if we go with another Node microservice app or in a JSON file within S3 if we go with a Lambda function.
- If we prefer to have a single place wjere transcriptions are processed rather one function per account (above approach), we can use the approach shown here, using `twilio` SDK. To interact with Twilio APIs we'll need the credentials for each account. Since we are already storing and bundling them with `hrm-service`, we can continue to use the same approach and maintain the same source of truth.
- I personally don't see any reason to keep this functionality within `hrm-service`, but I'm not against it either.
- The next step for this is to decide if we prefer to have this "extract transcripts" functionality as AWS Lambda function or in a Node server. If we'll store the transcriptions in a Postgres DB the later one might be better, and have in mind that fetching each transcription from Twilio might take some time so Lambda functions can end up taking a long execution time (if the record sets are big).
- We'll need to decide what to do in case of errors while trying to perform this operation in the case of:
  - Failure when trying to request the target transcripts to `hrm-service`.
  - Failure when trying to fetch the transcript from Twilio system (bad credentials, network blip, Twilio outage). This can occur for the whole record set or for some of them.
  - Failure when trying to store the transcripts in our storage.

### Checklist
- [x] [Corresponding issue has been opened](https://bugs.benetech.org/browse/CHI-1100)
- [ ] New tests added

### Verification steps
- Compile and run the local server.
- Add contact records with real Twilio chat channels to the DB (insert statements for this can be found in `/transcripts/test-transcripts.js`).
- Run `POSTGRES_PORT=5433 node dist/transcripts/test-transcripts.js` to execute the extract transcript functionality.
- Check that the files with the transcripts are stored locally and maybe confirm that the messages match the ones in Twilio system.

### CODE REVIEW

File: hrm-service/src/contact/sql/contact-get-columns-for-transcripts-sql.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    AND ("channelSid" != '') IS TRUE
+    AND (
+      $<dateFrom> IS NULL OR "createdAt" >= $<dateFrom>
+      -- $<dateFrom> IS NULL OR "timeOfContact" >= $<dateFrom>
+    )
+    AND (
+      $<dateTo> IS NULL OR "createdAt" <= $<dateTo>
+      -- $<dateTo> IS NULL OR "timeOfContact" <= $<dateTo>
+    );
+`;
```
Comment (GPaoloni) [inline hrm-service/src/contact/sql/contact-get-columns-for-transcripts-sql.ts:19]: About this query, two comments:
- `(stringStatement != '') IS TRUE` is a way to check emptiness for strings (not NULL nor empty string).
- I think we want to use `createdAt` here but let me know if you think `timeOfContact` is better for this. `timeOfContact` can only differ for offline contacts tho, and here we only care about chat contacts.

**Comment by nickhurlburt** [comment]
This looks good as a POC -- not sure we should merge it as I don't know if we want this in the production system. There are several things to address as a post-POC follow-up, such as batching queries and writes, as well as figuring out where to put everything. I recently learned that we can have [read replicas](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_PostgreSQL.Replication.ReadReplicas.html) on our postgres instances, which might be a good short-term way to do bulk queries.

**Comment by stephenhand** [comment]
Looks great @GPaoloni - much more confident of the processing approach now you've done this!

@GPaoloni @nickhurlburt

For an architecture going forward, I was wondering if it was worth considering something more generic for processing async jobs associated with contacts?

Now this might be a little OTT for solving this one problem, but I'm thinking as a generic batch async job processing framework, because we already have similar problems that we rolled specific solutions for (Dual Write comes to mind, maybe even CSAM reports? maybe CSAM is a bit different)

I was thinking we could have a separate table called 'ContactJobs' with something like the following fields: accountSid, contactId, jobType, taskSid, channelSid, lastAttempted, timeCompleted, completedStatus (success, error, w/e)

New contacts could add a row to the table for each 'job' needed (taskType = 'exportConversation' or taskType = 'dualWrite' etc.)

Then we could have an 'out' SQS queue for each accountSid/jobType. A process running in HRM on a regular interval could select incomplete jobs from the table and post job messages onto the queue, updating the 'lastAttempted' date (this field could be used to enforce a maximum attempt frequency for each job type)

On the other end of the queue is an SNS activated lambda that actions the job (possibly calling a Twilio Serverless function if it needs to be executed in 'Twilio Land') and on completion post the result to a second 'in' SQS queue

The same regular HRM process that dispatches jobs would also pull results off this queue and update the ContactJobs table (would probably do this step first)

```
|-----------|    |-----|                       
|ContactJobs| => |     | ====== SQS 'out' ====> |------------|
| Database  |    | HRM | (per account/job type) | Job Lambda |
|   Table   | <= |     | <===== SQS 'in' ====== |------------|
|-----------|    |-----|
```

Benefits:

* Provides separation of concerns between async processing jobs and HRM's core purpose
* Separate table reduces contention on Contacts, freeing it up to serve data for non job related requests.
* The async nature of the job state processing means the job lambda could be called directly from elsewhere, and the job state all still gets handled
* A separate table means that the job could even be completed and that status be recorded BEFORE the contact record gets written (linked using the taskSid or channelSid before the contactId exists)
* Relying on regular batch processing and queues means we have a lot of control over the rate we process any async jobs (it's down to the lambda whether we process jobs in parallel or in series etc.)
* Using SQS retains state of completed jobs if HRM is down

We could possibly simplify this a little by broadcasting SNS events to a lambda from HRM directly and only having an 'in' queue, but the 'out' queue does allow us more flexibility & control over processing rates.

Anyway, something for people to think about - if it's not a complete non starter I'll write it up somewhere better than a PR comment! 😂

**Comment by GPaoloni** [comment]
@stephenhand I have to step up my async queues skills, but I get the idea and really like it.

**Comment by nickhurlburt** [comment]
We are definitely moving closer to needing to have more async processing. This is a good starter proposal for that. I could bikeshed all day on the details, but I will resist :)  I could go further and say it would be great to separate async resources from our synchronous production system altogether, though I have mixed feelings about the resulting DevOps overhead. 

We may also want to give a little thought to other async needs we will have -- such as data warehouse exporting, helpline data pulls, and that insights correctness job that we've left broken for months -- to see if we want to work any of that in. That also may be too much at once.

**Review by stephenhand — COMMENTED** [review]
.

**Pull request closed without merging**